Add "placeholder" (pre-MVP) implementation of osu!catch editor

ekrctb opened the pull request:
It is a set of files that are needed for any plan of the catch editor anyway.
Due to complexity, juice stream editing/placement is not implemented in this PR.

### Known bugs (I expect this PR to be merged with known issues)

- Catcher size is not changed when the circle size is changed
- Scrolling speed is not changed when the approach rate is changed
- Objects can go out of bounds (only tiny droplets are confined to the playfield)
- Catcher trails/hit lighting are duplicated. Catcher animation is not rewound correctly.
- Fruit outline is always a circle even with a non-default skin.

Fixes to some issues are not included in this PR because that requires changes to the `osu.Game` code or non-editor parts.
Others are not implemented yet due to complexity.

### Not implemented (features needed even for basic editing)

- Editing/placement of juice streams
  - At least piecewise-linear ones.
- Position snapping
- Copy/paste (serialization)
- Probably a lot more
- Hyper state of hit objects should be updated real-time when position is changed <https://github.com/ppy/osu/pull/13617#issuecomment-865827951>
- Hit object shouldn't be expired when caught by the catcher <https://github.com/ppy/osu/pull/13617#issuecomment-866247870>



https://user-images.githubusercontent.com/191335/124296661-2e665300-db95-11eb-8da5-ad34241a6e19.mp4

ekrctb committed:
Fix NRE when hit object blueprint is not implemented

ekrctb committed:
Add empty catch hit object composer

ekrctb committed:
Add basic selection blueprint movement logic

ekrctb committed:
Add basic juice stream selection blueprint

ekrctb committed:
Add outline to selected fruit

ekrctb committed:
Add basic fruit placement tool

ekrctb committed:
Add basic banana shower placement tool

ekrctb committed:
Disable caught object stacking in editor

peppy commented on osu.Game.Rulesets.Catch/Edit/Blueprints/Components/FruitOutline.cs:
```diff
@@ -0,0 +1,39 @@
+            Anchor = Anchor.BottomLeft;
+            Origin = Anchor.Centre;
+            Size = new Vector2(2 * CatchHitObject.OBJECT_RADIUS);
+            // TODO: use skinned component?
```
Just for reference, in general we're aiming to keep the editor blueprints as design/skin agnostic. They should exist as a layer that is placed over the top of the actual gameplay, and can be toggled at any point in time. So what you have is probably more the direction we'd be going, like in other rulesets' editor implementations.

peppy commented:
One other issue to add to the list is that hyperdash state doesn't update on `X` movement of fruits (only `Y` / `Time` changes trigger an update currently).

smoogipoo reviewed (dismissed):
Looks fine to me as an initial pass :)

bdach commented on osu.Game.Rulesets.Catch/Edit/BananaShowerCompositionTool.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace osu.Game.Rulesets.Catch.Edit
+{
+    public class BananaShowerCompositionTool : HitObjectCompositionTool
```
May also want to override `CreateIcon()` - other rulesets do

```csharp
public override Drawable CreateIcon() => new BeatmapStatisticIcon(BeatmapStatisticsIconType.Spinners);
```

bdach commented on osu.Game.Rulesets.Catch/Edit/FruitCompositionTool.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace osu.Game.Rulesets.Catch.Edit
+{
+    public class FruitCompositionTool : HitObjectCompositionTool
```
similarly as above here

```csharp
public override Drawable CreateIcon() => new BeatmapStatisticIcon(BeatmapStatisticsIconType.Circles);
```

bdach commented on osu.Game/Screens/Edit/Compose/Components/ComposeBlueprintContainer.cs:
```diff
@@ -256,7 +256,7 @@ protected sealed override SelectionBlueprint<HitObject> CreateBlueprintFor(HitOb
+            return CreateHitObjectBlueprintFor(item)?.With(b => b.DrawableObject = drawable);
         }
 
         public virtual HitObjectSelectionBlueprint CreateHitObjectBlueprintFor(HitObject hitObject) => null;
```
may as well mark as `[CanBeNull]`, if we're doing the change above

bdach reviewed (commented):
A few minor things that may as well be fixed here.

bdach commented:
Another issue for the pile: Timeline blueprints pop out of existence shortly after their corresponding objects are caught in the playfield, unless they are currently selected. (At least they pop back up on rewind, though.)

ekrctb committed:
Add icon for composition tools

ekrctb committed:
Remove comment about using skin for blueprint

As the current game-wise direction is not using skin elements in blueprints.
The design of the blueprint could be improved somehow, though.

ekrctb committed:
Add `[CanBeNull]` to methods returning null by default

ekrctb committed:
Order field and properties consistently

ekrctb committed:
Use more consistent method names

ekrctb committed:
Add some comments about logic

ekrctb committed:
Fix banana shower placement outline initial opacity

ekrctb committed:
Fix "possible NRE" inspection

peppy commented:
This looks great as a first implementation.

One more thing which should be considered a relatively high priority TODO is adding a time snap grid (basically bar lines on the playfield). Could be implemented similarly to mania's, I think.

![20210623 135840 (dotnet)](https://user-images.githubusercontent.com/191335/123038335-24bd4c80-d42b-11eb-967b-a79a3361561a.gif)
